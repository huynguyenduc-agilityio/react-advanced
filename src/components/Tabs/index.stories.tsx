import type { Meta, StoryObj } from '@storybook/react';
import { Flex } from '@chakra-ui/react';

// Icons
import { HiMiniPencil } from 'react-icons/hi2';
import { FaCreditCard, FaUser } from 'react-icons/fa';
import { RiNotification2Fill } from 'react-icons/ri';

// Components
import Tabs from '.';

const TAB_LIST = [
  {
    label: 'Personal Information',
    icon: HiMiniPencil,
    content: (
      <p>
        1. Lorem ipsum dolor sit, amet consectetur adipisicing elit. Magnam, ut
        odio. Enim sint excepturi dolor vero provident, quisquam et totam
        doloribus aspernatur, laborum molestias, dolores ipsum non nobis labore
        temporibus.
      </p>
    ),
  },
  {
    label: 'Team',
    icon: FaUser,
    content: (
      <p>
        2. Lorem ipsum dolor sit amet consectetur adipisicing elit. Aut enim
        dolorum similique corporis rem sequi, eos numquam quia velit vel non.
        Suscipit non fugiat nesciunt exercitationem totam, ipsa molestias
        praesentium.
      </p>
    ),
  },
  {
    label: 'Billing',
    icon: FaCreditCard,
    content: (
      <p>
        3. Lorem ipsum dolor sit amet consectetur adipisicing elit. Tempore
        incidunt consequuntur modi sint sunt, maiores ullam est pariatur impedit
        libero labore nisi provident perspiciatis. Autem, nulla eligendi! Rerum,
        itaque veritatis!
      </p>
    ),
  },
  {
    label: 'Notifications',
    icon: RiNotification2Fill,
    content: (
      <p>
        4. Lorem ipsum dolor sit amet consectetur adipisicing elit. Tempore
        doloremque quos quod, ut deleniti saepe excepturi quibusdam! Voluptatum
        eius, voluptatibus animi cum illo, adipisci reiciendis, nam sit maiores
        a praesentium?
      </p>
    ),
  },
];

const defaultProps = {
  tabs: TAB_LIST,
};

const meta = {
  title: 'Components/Tabs',
  component: Tabs,
  parameters: {
    layout: 'centered',
  },

  tags: ['autodocs'],
  render: (args) => (
    <Flex
      justifyContent="center"
      w="1000px"
      p={9}
      bgColor="spaceBlue"
      color="pastelBlue"
    >
      <Tabs {...args} />
    </Flex>
  ),
} satisfies Meta<typeof Tabs>;

export default meta;

type Story = StoryObj<typeof Tabs>;

export const Default: Story = {
  args: defaultProps,
};
