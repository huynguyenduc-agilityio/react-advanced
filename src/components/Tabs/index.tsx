import { ReactNode } from 'react';
import { IconType } from 'react-icons';
import {
  Tabs as ChakraTabs,
  Icon,
  Tab,
  TabList,
  TabPanel,
  TabPanels,
  Text,
} from '@chakra-ui/react';

// Themes
import { fonts } from '@/themes/bases/typography';

type ITabItem = {
  label: string;
  content: ReactNode;
  icon: IconType;
};

type ITabsProps = {
  tabs: ITabItem[];
};

const Tabs = ({ tabs }: ITabsProps) => {
  return (
    <ChakraTabs orientation="vertical" variant="unstyled">
      <TabList minW="250px" gap={2} mr="52px">
        <Text
          as="header"
          fontFamily={fonts.heading}
          fontSize="lg"
          mb={4}
          fontWeight={500}
          color="white"
        >
          Credentials
        </Text>
        {tabs.map(({ label, icon }) => (
          <Tab
            key={label}
            justifyContent="flex-start"
            py={3}
            px={3}
            borderRadius="lg"
            _selected={{ color: 'white', bg: 'darkBlue' }}
            _hover={{ bg: 'darkBlue' }}
          >
            <Icon as={icon} boxSize={3} />
            <Text fontSize="md" ml={1}>
              {label}
            </Text>
          </Tab>
        ))}
      </TabList>
      <TabPanels borderLeft="1px solid" borderColor="slateBlue">
        {tabs.map(({ label, content }) => (
          <TabPanel p="0" ml="97px" key={label}>
            {content}
          </TabPanel>
        ))}
      </TabPanels>
    </ChakraTabs>
  );
};

export default Tabs;
