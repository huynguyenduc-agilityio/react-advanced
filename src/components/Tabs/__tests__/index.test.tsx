import { render } from '@testing-library/react';
import { FaCreditCard, FaUser } from 'react-icons/fa';
import { HiMiniPencil } from 'react-icons/hi2';
import { RiNotification2Fill } from 'react-icons/ri';

import Tabs from '..';

const mockTabs = [
  {
    label: 'Personal Information',
    icon: HiMiniPencil,
    content: <></>,
  },
  {
    label: 'Team',
    icon: FaUser,
    content: <></>,
  },
  {
    label: 'Billing',
    icon: FaCreditCard,
    content: <></>,
  },
  {
    label: 'Notifications',
    icon: RiNotification2Fill,
    content: <></>,
  },
];

describe('Tabs component', () => {
  it('renders correctly and matches snapshot (only tabs structure)', () => {
    const { container } = render(<Tabs tabs={mockTabs} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
