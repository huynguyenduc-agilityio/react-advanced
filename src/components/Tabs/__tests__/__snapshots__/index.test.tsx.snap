// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component should match snapshot 1`] = `
<div
  class="chakra-tabs css-79elbk"
>
  <div
    aria-orientation="vertical"
    class="chakra-tabs__tablist css-4pg0xm"
    role="tablist"
  >
    <header
      class="chakra-text css-19fkgi9"
    >
      Credentials
    </header>
    <button
      aria-controls="tabs-:r0:--tabpanel-0"
      aria-disabled="false"
      aria-selected="true"
      class="chakra-tabs__tab css-1cn99nz"
      data-index="0"
      id="tabs-:r0:--tab-0"
      role="tab"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-zdliob"
        fill="currentColor"
        focusable="false"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m2.695 14.762-1.262 3.155a.5.5 0 0 0 .65.65l3.155-1.262a4 4 0 0 0 1.343-.886L17.5 5.501a2.121 2.121 0 0 0-3-3L3.58 13.419a4 4 0 0 0-.885 1.343Z"
        />
      </svg>
      <p
        class="chakra-text css-1mitela"
      >
        Personal Information
      </p>
    </button>
    <button
      aria-controls="tabs-:r0:--tabpanel-1"
      aria-disabled="false"
      aria-selected="false"
      class="chakra-tabs__tab css-1cn99nz"
      data-index="1"
      id="tabs-:r0:--tab-1"
      role="tab"
      tabindex="-1"
      type="button"
    >
      <svg
        class="chakra-icon css-zdliob"
        fill="currentColor"
        focusable="false"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
        />
      </svg>
      <p
        class="chakra-text css-1mitela"
      >
        Team
      </p>
    </button>
    <button
      aria-controls="tabs-:r0:--tabpanel-2"
      aria-disabled="false"
      aria-selected="false"
      class="chakra-tabs__tab css-1cn99nz"
      data-index="2"
      id="tabs-:r0:--tab-2"
      role="tab"
      tabindex="-1"
      type="button"
    >
      <svg
        class="chakra-icon css-zdliob"
        fill="currentColor"
        focusable="false"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 576 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 432c0 26.5 21.5 48 48 48h480c26.5 0 48-21.5 48-48V256H0v176zm192-68c0-6.6 5.4-12 12-12h136c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H204c-6.6 0-12-5.4-12-12v-40zm-128 0c0-6.6 5.4-12 12-12h72c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zM576 80v48H0V80c0-26.5 21.5-48 48-48h480c26.5 0 48 21.5 48 48z"
        />
      </svg>
      <p
        class="chakra-text css-1mitela"
      >
        Billing
      </p>
    </button>
    <button
      aria-controls="tabs-:r0:--tabpanel-3"
      aria-disabled="false"
      aria-selected="false"
      class="chakra-tabs__tab css-1cn99nz"
      data-index="3"
      id="tabs-:r0:--tab-3"
      role="tab"
      tabindex="-1"
      type="button"
    >
      <svg
        class="chakra-icon css-zdliob"
        fill="currentColor"
        focusable="false"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M22 20H2V18H3V11.0314C3 6.04348 7.02944 2 12 2C16.9706 2 21 6.04348 21 11.0314V18H22V20ZM9.5 21H14.5C14.5 22.3807 13.3807 23.5 12 23.5C10.6193 23.5 9.5 22.3807 9.5 21Z"
        />
      </svg>
      <p
        class="chakra-text css-1mitela"
      >
        Notifications
      </p>
    </button>
  </div>
  <div
    class="chakra-tabs__tab-panels css-m0szyg"
  >
    <div
      aria-labelledby="tabs-:r0:--tab-0"
      class="chakra-tabs__tab-panel css-rk1yrb"
      id="tabs-:r0:--tabpanel-0"
      role="tabpanel"
      tabindex="0"
    />
    <div
      aria-labelledby="tabs-:r0:--tab-1"
      class="chakra-tabs__tab-panel css-rk1yrb"
      hidden=""
      id="tabs-:r0:--tabpanel-1"
      role="tabpanel"
      tabindex="0"
    />
    <div
      aria-labelledby="tabs-:r0:--tab-2"
      class="chakra-tabs__tab-panel css-rk1yrb"
      hidden=""
      id="tabs-:r0:--tabpanel-2"
      role="tabpanel"
      tabindex="0"
    />
    <div
      aria-labelledby="tabs-:r0:--tab-3"
      class="chakra-tabs__tab-panel css-rk1yrb"
      hidden=""
      id="tabs-:r0:--tabpanel-3"
      role="tabpanel"
      tabindex="0"
    />
  </div>
</div>
`;
